gcapi/gcapi.py

```diff
@@ -353,115 +337,145 @@ class RetinaETDRSGridAnnotationsAPI(APIBase, ModifiableMixin):
+        user_upload = self.create(filename=filename)
 
-        Raises
-        ------
-        HTTPError
-            Raised if a chunk cannot be uploaded.
-        """
-        if not content:
-            content = load_input_data(filename)
-        upload_id = generate_new_upload_id(content)
-        start_byte = 0
-        content_io = BytesIO(content)
-        max_chunk_length = 2 ** 23
-        results = []
+        pk = user_upload["pk"]
+        s3_upload_id = user_upload["s3_upload_id"]
+
+        try:
+            parts = self._put_fileobj(
+                fileobj=fileobj, pk=pk, s3_upload_id=s3_upload_id
+            )
+        except Exception:
+            self.abort_multipart_upload(pk=pk, s3_upload_id=s3_upload_id)
+            raise
+
+        return self.complete_multipart_upload(
+            pk=pk, s3_upload_id=s3_upload_id, parts=parts
+        )
+
+    def _put_fileobj(self, *, fileobj, pk, s3_upload_id):
+        part_number = 1  # s3 uses 1-indexed chunks
+        presigned_urls = {}
+        parts = []
 
         while True:
-            chunk = content_io.read(max_chunk_length)
+            chunk = fileobj.read(self.chunk_size)
 
             if not chunk:
                 break
 
-            end_byte = start_byte + len(chunk)
-
-            try:
-                result = self._upload_file_with_exponential_backoff(
-                    {
-                        "start_byte": start_byte,
-                        "end_byte": end_byte,
-                        "chunk": chunk,
-                        "content": content,
-                        "upload_id": upload_id,
-                        "filename": str(filename),
-                    }
+            if str(part_number) not in presigned_urls:
+                presigned_urls.update(
+                    self._get_next_presigned_urls(
+                        pk=pk,
+                        s3_upload_id=s3_upload_id,
+                        part_number=part_number,
+                    )
                 )
-            except HTTPError as e:
-                raise e
 
-            results.append(result)
+            response = self._put_chunk(
+                chunk=chunk, url=presigned_urls[str(part_number)]
+            )
+
+            parts.append(
+                {"ETag": response.headers["ETag"], "PartNumber": part_number}
+            )
+
+        return parts
+
+    def _get_next_presigned_urls(self, *, pk, s3_upload_id, part_number):
+        response = self.generate_presigned_urls(
+            pk=pk,
+            s3_upload_id=s3_upload_id,
+            part_numbers=[
+                *range(part_number, part_number + self.n_presigned_urls)
+            ],
+        )
+        return response["presigned_urls"]
+
+    def _put_chunk(self, *, chunk, url):
+        num_retries = 0
+        e = Exception
 
-            start_byte += len(chunk)
+        while num_retries < self.max_retries:
+            try:
+                result = self._client.request(
```

Comment:
Maybe good to add a test to check that the token is not sent along with this request. When you only see this code, the difference between `self._client(...)` and `self._client.request(...)` is unclear so I can imagine that someone switch these up by accident.

Reply:
Done using httpbin, not sure that this is the best method.